docs/repo/property-pages/how-to-add-a-new-launch-profile-kind.md

```diff
@@ -189,3 +202,107 @@ public sealed class MyPackage : AsyncPackage
 ```
 
 And you're done. This Launch Profile UI will be available in any project with matching capabilities.
+
+## Option 3: `IRuleObjectProvider`
+
+_Note this option is only available starting in VS 2022 Update 1 (Dev17.1)._
+
+In this option, `Rule` objects are defined in code and provided to the project system via MEF-exported implementations of the `IRuleObjectProvider` type. This will be an attractive option if you already have an assembly participating in the VS MEF composition, don't have a .props or .targets file to hold `PropertyPageSchema` items, or prefer to define your `Rule`s in code rather than XAML.
+
+This is also the simplest option if you need to localize your launch profile, as you can read the localized text from .resx files or similar rather than translating the entire .xaml file and distributing the translations.
+
+### Step 1: Add the Project System SDK
+
+Use the NuGet Package Manager to add the Microsoft.VisualStudio.ProjectSystem.Sdk package package to your project.
+
+### Step 2: Define the `IRuleObjectProvider`
+
+Create a new class with the following contents:
+
+```csharp
+using System;
+using System.Collections.Generic;
+using Microsoft.Build.Framework.XamlTypes;
+using Microsoft.VisualStudio.ProjectSystem;
+using Microsoft.VisualStudio.ProjectSystem.Properties;
+
+namespace RuleObjectProviderDemo
+{
+    [ExportRuleObjectProvider(name: "MyRuleObjectProvider", context: "Project")]
+    [AppliesTo("MyUniqueProjectCapability")]
+    [Order(0)]
+    public class MyRuleObjectProvider : IRuleObjectProvider
+    {
+        private List<Rule> rules;
+
+        public IReadOnlyCollection<Rule> GetRules()
```

Comment:
Does the implementation need to be thread-safe? Ditto below.

Reply:
Good idea.